### DISCUSSION TITLE
Querying against a postgresql ARRAY column checking for membership

### DISCUSSION THREAD

**Discussion by Kaelten** [sqlalchemy/sqlalchemy#12055] [Usage Questions]
I have a model that has a column that looks like this:

```python
eating_occasions: Mapped[tuple[EatingOccasion] | None]
```

`EatingOccasion` is a StrEnum descendent via the below subclass to handle the mappings.

```python
class GraphqlEnum(enum.StrEnum):
    """Specialized StrEnum subclass that handles type map registration with SqlAlchemy."""

    def __init_subclass__(cls) -> None:
        SQLModel.registry.update_type_annotation_map(
            {
                list[cls]: ARRAY(Enum(cls, name='graphql_enum')),  # type: ignore[valid-type]
                tuple[cls]: ARRAY(Enum(cls, name='graphql_enum'), as_tuple=True),  # type: ignore[valid-type]
            },
        )
```

When I'm trying to query against the `eating_occasions` side is where I'm getting issues though.  I'm attempting to query rows where the array contains a specific value.  I'm not sure if I'm just dense and am having a hard time finding the right pattern or what, but below is the various things I've tried and while I've identified a few options that work they all seem to have gotchas associated.  Please help me understand the right path forward. :) 

### .contains
```python
RecipeSearchIndex.eating_occasions.contains(EatingOccasion.breakfast)
```

This results in an invalid query being generated
```
WHERE recipe_search_index.eating_occasions @> %s::graphql_enum[]
2024-10-31 13:52:53,001 INFO sqlalchemy.engine.Engine [generated in 0.00020s] (['b', 'r', 'e', 'a', 'k', 'f', 'a', 's', 't'],)
```

### .any
```python
RecipeSearchIndex.eating_occasions.any(EatingOccasion.breakfast)
```

This results in a valid query, but gives a mypy error and is identified as deprecated
```
error: Argument 1 to "any" of "PropComparator" has incompatible type "EatingOccasion"; expected "ColumnElement[bool] | _HasClauseElement[bool] | SQLCoreOperations[bool] | ExpressionElementRole[bool] | TypedColumnsClauseRole[bool] | Callable[[], ColumnElement[bool]] | LambdaElement | None"  [arg-type]
```

### .any_
```python
RecipeSearchIndex.eating_occasions.any_() == EatingOccasion.breakfast
```

I thought this was the right way forward based on docs, but I get this error:
```
sqlalchemy.exc.StatementError: (builtins.TypeError) Not a boolean value: <EatingOccasion.breakfast: 'breakfast'>
[SQL: SELECT ...
FROM recipe_search_index 
WHERE %s::BOOLEAN = ANY (recipe_search_index.eating_occasions)]
[parameters: [{}]]
```

I also tested inverting the order, but the same results happen.

### .contains([])
```python
RecipeSearchIndex.eating_occasions.contains([EatingOccasion.breakfast])
```

This seems to return the results expected, but appears to be using list subset mechanics on the query side?  I'm also concerned since it'd be very trivial to forget to wrap the args in a list.

**Comment by CaselIT** [comment]
Hi,

Contains accepts an array, so the correct usecase is the last one

**Reply by Kaelten** [reply]
looking at the contains query it's generating this:

```
WHERE recipe_search_index.eating_occasions @> %s::graphql_enum[]
```

However, what I was trying to generate is:

```
WHERE %s::graphql_enum = ANY(recipe_search_index.eating_occasions)
```

These two query styles do generate different query plans and will use indexes differently. Thus some of my concern outside of the fact that passing in the enum value without wrapping it in the list works fine from a type checking and is error free while executing (just without any results).

It's also quiet confusing that the `.any()_` doesn't work as it appears that's generating the second style of sql, but is trying to cast it into a boolean for some reason.

**Reply by Kaelten** [reply]
Thought: would it be a reasonable request that `.contains` becomes a bit smarter in the future?  i.e. if it's passed a non-sequence to generate the `ANY()` syntax and if it's passed a sequence to generate the current syntax?

**Reply by CaselIT** [reply]
> However, what I was trying to generate is:
> 
> ```
> WHERE %s::graphql_enum = ANY(recipe_search_index.eating_occasions)
> ```

You can write that explicitely:

```py
EatingOccasion.breakfast == sa.any_(RecipeSearchIndex.eating_occasions)
```

> Thought: would it be a reasonable request that `.contains` becomes a bit smarter in the future? i.e. if it's passed a non-sequence to generate the `ANY()` syntax and if it's passed a sequence to generate the current syntax?

That would not be a good api. Contains should render the contains operator: https://www.postgresql.org/docs/current/functions-array.html

There are multiple methods to render any:
- legacy: https://docs.sqlalchemy.org/en/20/core/type_basics.html#sqlalchemy.types.ARRAY.Comparator.any
- current: https://docs.sqlalchemy.org/en/20/core/sqlelement.html#sqlalchemy.sql.expression.ColumnOperators.any_

Both should work
```py
RecipeSearchIndex.eating_occasions.any(EatingOccasion.breakfast)
EatingOccasion.breakfast == RecipeSearchIndex.eating_occasions.any_()
```

**Reply by Kaelten** [reply]
> You can write that explicitely:
> 
> ```python
> EatingOccasion.breakfast == sa.any_(RecipeSearchIndex.eating_occasions)
> ```

As mentioned, I tried that and SqlAlchemy errors.
```
sqlalchemy.exc.StatementError: (builtins.TypeError) Not a boolean value: <EatingOccasion.breakfast: 'breakfast'>
[SQL: SELECT ...
FROM recipe_search_index 
WHERE %s::BOOLEAN = ANY (recipe_search_index.eating_occasions)]
[parameters: [{}]]
```

> > Thought: would it be a reasonable request that `.contains` becomes a bit smarter in the future? i.e. if it's passed a non-sequence to generate the `ANY()` syntax and if it's passed a sequence to generate the current syntax?
> 
> That would not be a good api. Contains should render the contains operator: https://www.postgresql.org/docs/current/functions-array.html

We could get into semantics about containing an array vs containing an element and the semantics you're describing don't line up with the standards in python for testing an array membership.  e.g.

```python
In [1]: 'breakfast' in ['breakfast', 'lunch']
Out[1]: True

In [2]: ['breakfast'] in ['breakfast', 'lunch']
Out[2]: False
```

The closer analog to what you're describing is how we normally handle checking sets against each other using `issubset` or the `<=` operator.   These are arguably pedantic distinctions, but if conforming to pythonic idioms is a concern, I'd suggest reconsidering this.

It's more concerning to me that neither the legacy nor new version of the any syntax work.

> There are multiple methods to render any:
> 
> * legacy: https://docs.sqlalchemy.org/en/20/core/type_basics.html#sqlalchemy.types.ARRAY.Comparator.any
> * current: https://docs.sqlalchemy.org/en/20/core/sqlelement.html#sqlalchemy.sql.expression.ColumnOperators.any_
> 
> Both should work
> 
> ```python
> RecipeSearchIndex.eating_occasions.any(EatingOccasion.breakfast)
> EatingOccasion.breakfast == RecipeSearchIndex.eating_occasions.any_()
> ```

As mentioned above the modern version errors and the legacy version works, but generates a mypy typing error.

**Reply by CaselIT** [reply]
> As mentioned, I tried that and SqlAlchemy errors.
> 
> ```
> sqlalchemy.exc.StatementError: (builtins.TypeError) Not a boolean value: <EatingOccasion.breakfast: 'breakfast'>
> [SQL: SELECT ...
> FROM recipe_search_index 
> WHERE %s::BOOLEAN = ANY (recipe_search_index.eating_occasions)]
> [parameters: [{}]]
> ```

This looks strange. What dialect is returning this error? Can you provide a runnable example that generate such errors (or at least sqlalchemy version, and dialect in question )? That may be a bug, if so it will be fixed so that you can indeed run that syntax correctly

> We could get into semantics about containing an array vs containing an element and the semantics you're describing don't line up with the standards in python for testing an array membership. e.g.
> 
> ```python
> In [1]: 'breakfast' in ['breakfast', 'lunch']
> Out[1]: True
> 
> In [2]: ['breakfast'] in ['breakfast', 'lunch']
> Out[2]: False
> ```

But you are not running python statements, you are generating sql. This is a custom type present only in postgresql. They defined the syntax for the contain operator on array. An user would expect that calling `contains` on a postgresql array will execute the contain array operator of postgresql.
